django_projects/locallibrary/catalog/tests/test_views.py

```diff
@@ -0,0 +1,249 @@
+        response = self.client.get(reverse('my-borrowed'))
+
+        self.assertEqual(str(response.context['user']), 'testuser1')
+        self.assertEqual(response.status_code, 200)
+
+        self.assertTemplateUsed(response, 'catalog/bookinstance_list_borrowed_user.html')
+
+    def test_only_borrowed_books_in_list(self):
+        login = self.client.login(username='testuser1', password='1X<ISRUkw+tuK')
+        response = self.client.get(reverse('my-borrowed'))
+
+        self.assertEqual(str(response.context['user']), 'testuser1')
+        self.assertEqual(response.status_code, 200)
+
+        self.assertTrue('bookinstance_list' in response.context)
+        self.assertEqual(len(response.context['bookinstance_list']), 0)
+
+        get_ten_books = BookInstance.objects.all()[:10]
+
+        for copy in get_ten_books:
+            copy.status = 'o'
+            copy.save()
+
+        response = self.client.get(reverse('my-borrowed'))
+        self.assertEqual(str(response.context['user']), 'testuser1')
+        self.assertEqual(response.status_code, 200)
+
+        self.assertTrue('bookinstance_list' in response.context)
+
+        for bookitem in response.context['bookinstance_list']:
+            self.assertEqual(response.context['user'], bookitem.borrower)
+            self.assertEqual(bookitem.status, 'o')
+
+    def test_pages_paginated_to_ten(self):
+        for copy in BookInstance.objects.all():
+            copy.status = 'o'
+            copy.save()
+
+        login = self.client.login(username='testuser1', password='1X<ISRUkw+tuK')
+        response = self.client.get(reverse('my-borrowed'))
+
+        self.assertEqual(str(response.context['user']), 'testuser1')
+        self.assertEqual(response.status_code, 200)
+
+        self.assertEqual(len(response.context['bookinstance_list']), 10)
+
+    def test_pages_ordered_by_due_date(self):
+        for copy in BookInstance.objects.all():
+            copy.status = 'o'
+            copy.save()
+
+        login = self.client.login(username='testuser1', password='1X<ISRUkw+tuK')
+        response = self.client.get(reverse('my-borrowed'))
+
+        self.assertEqual(str(response.context['user']), 'testuser1')
+        self.assertEqual(response.status_code, 200)
+
+        self.assertEqual(len(response.context['bookinstance_list']), 10)
+
+        last_date = 0
+        for copy in response.context['bookinstance_list']:
+            if last_date == 0:
+                last_date = copy.due_back
+            else:
+                self.assertTrue(last_date <= copy.due_back)
+
+class RenewBookInstancesViewTest(TestCase):
+    def setUp(self):
+        test_user1 = User.objects.create_user(username='testuser1', password='1X<ISRUkw+tuK')
+        test_user1.save()
+
+        test_user2 = User.objects.create_user(username='testuser2', password='2HJ1vRV0Z&3iD')
+        test_user2.save()
+        permission = Permission.objects.get(name='Set book as returned')
```

Comment:
em tạo data Permission đã nhé, DB của môi trường test độc lập với môi trường development nha e
```suggestion
        permission = Permission.objects.(name='Set book as returned', ...)
```

Reply:
![image](https://user-images.githubusercontent.com/62171695/121497340-892ce480-ca05-11eb-9344-4bfae9d6bbe0.png)
Sau khi đổi thành create thì nó lỗi ở database ạ.

Reply:
Em phải tạo dữ liệu đầy đủ nha. Tham khảo ví dụ 
https://docs.djangoproject.com/en/3.2/topics/auth/default/#programmatically-creating-permissions
`from myapp.models import BlogPost
from django.contrib.auth.models import Permission
from django.contrib.contenttypes.models import ContentType

content_type = ContentType.objects.get_for_model(BlogPost)
permission = Permission.objects.create(
    codename='can_publish',
    name='Can Publish Posts',
    content_type=content_type,
)`

Reply:
![image](https://user-images.githubusercontent.com/62171695/121500779-cfd00e00-ca08-11eb-8c3a-c8b8bb44a77a.png)
Test OK rồi chị ạ.